### ISSUE TITLE
Progressive Registration

### ISSUE THREAD

**Issue by maxceem** [appirio-tech/connect-app#3833]
This is a new big feature as per design https://marvelapp.com/e0gjig7/screen/67320435

![image](https://user-images.githubusercontent.com/146016/78241669-735e8e00-74e9-11ea-8466-193e5a2bcb4e.png)

**Comment by maxceem** [comment]
@vikasrohit the design for this form introduces the 3rd variation of the design for this form. We already have a **registration form**:

![image](https://user-images.githubusercontent.com/146016/78241809-ac96fe00-74e9-11ea-8052-4d5b3a494703.png)

and **Profile Information Form**:

![image](https://user-images.githubusercontent.com/146016/78241881-c59faf00-74e9-11ea-9286-a7881e085e31.png)

I suggest reusing the design from the **Profile Information Form** for this new form instead of creating a new variation of design:

![image](https://user-images.githubusercontent.com/146016/78241936-e0722380-74e9-11ea-85e0-fe0970bbc0d2.png)

**Comment by vikasrohit** [comment]
I am okay with it for now as I don't see much difference in new design as well. However, please do keep the required vs optional fields requirements as is i.e. use the requirement doc as primary source for making fields required or optional (based on roles).

**Comment by coderhacker** [comment]
@maxceem 
QA testing against screen https://marvelapp.com/e0gjig7/screen/67320439:

**1.**  Back arrow missing
**2.** Header Background is not matched with the provided design.
**3.** Header text **Profile Information** is missing.
**4.** Paragraph text "You have the incomplete required profile information. Please complete your profile information before able to submit your project request." is missing.
**5 to 12.** All labels are not matching with the provided design.
**13.** Cancel and Sent My Request button is missing
**14**
 ![image](https://user-images.githubusercontent.com/62358241/79827822-fa9e7380-83bc-11ea-8daf-4d7d136c4612.png) is missing.
**15.**
![image](https://user-images.githubusercontent.com/62358241/79828103-acd63b00-83bd-11ea-878e-6cb8014546b3.png) is missing.

![image](https://user-images.githubusercontent.com/62358241/79827988-5c5edd80-83bd-11ea-84f7-361c09985b16.png)

**PS**: All validation is working as expected.

CC: @codejamtc  @drasticdpk

**Comment by coderhacker** [comment]
Query: we need access for below google sheet. https://docs.google.com/document/d/1MwaHn_66uz5e6_1zXBTIjd6DqywZBWKHBzZwijFZM6M/edit#heading=h.tc9z5mq8parq

**Comment by maxceem** [comment]
@coderhacker we didn't have a task to update the Profile Information form. It should stay same as it is now.

Could you please, ask someone from QA team to share with you the google document with the task description.

**Comment by coderhacker** [comment]
@maxceem: I have updated the above comments. There is some minor issue which needs to fix.

**Comment by maxceem** [comment]
Thanks, @coderhacker.

> 5 to 12. All labels are not matching with the provided design.

We reuse our existing form design. So we keep labels on the left, not on the top as on the new design.

> 13. Cancel and Sent My Request button is missing

`Cancel` button was renamed to `Back` as we use the `Back` button on all other places. It was confirmed in the Google Document comments (unfortunately, comments are not included in the copy of the document I've shared with you, but there are not many comments).

`Sent My Request` button is there:
![image](https://user-images.githubusercontent.com/146016/79844176-fe81c400-83c3-11ea-9dff-d9b5635efc6c.png)

14. Good catch, have to fix it.

**Comment by maxceem** [comment]
14. is fixed now:

      ![image](https://user-images.githubusercontent.com/146016/79872424-706f0300-83ee-11ea-8e41-566da003a5a1.png)

     - note, that we are using our current standard link color and the size of the text is the same as we use on the other page https://connect.topcoder-dev.com/new-project (at the bottom).

@coderhacker could you please let me know if anything else has to be fixed per the clarifications above https://github.com/appirio-tech/connect-app/issues/3833#issuecomment-617037706

**Comment by maxceem** [comment]
We've found some issue with progressive registration when using a new empty Topcoder/Copilot account. So please, hold on with continue testing, we would fix it first.

cc @codejamtc

**Comment by codejamtc** [comment]
@maxceem I can see few issues on the form {Client Workflow}: ❌ 
https://monosnap.com/file/Ivl770NJ9SvMXu4fQO0tSiw0ugJHox

Reference: https://marvelapp.com/e0gjig7/screen/67320436 {Design} | https://docs.google.com/document/d/1M7IXe6Pvz4hEAsKy503_Fdbs6cLq0cYXUqxuWIBdPjI/edit# {Requirement}

1. `Company Name` is missing on the form. Existing form `My Profile > Profile Information` has the `Company Name` field https://monosnap.com/file/Ic7MGQ7QjdnCDPVngnE1KHZF5hfrAs. New design ref: https://marvelapp.com/e0gjig7/screen/67320436

![image](https://user-images.githubusercontent.com/42398485/80010442-0a3abc80-8490-11ea-9383-1ecc932b2699.png)

2. Since we have validations activated `Title`, `Business Email` and `Business Phone` labels are not in-line with the text fields

![image](https://user-images.githubusercontent.com/42398485/80010870-aebcfe80-8490-11ea-9def-4d8c84d19393.png)

3. `Send My Request` button is enabled even though the user filled an invalid email to the `Business Email` field

![image](https://user-images.githubusercontent.com/42398485/80011321-52a6aa00-8491-11ea-969e-31f59705acf5.png)

4. `Send my Request` button text should be `Send My Request` button
Ref: https://docs.google.com/document/d/1M7IXe6Pvz4hEAsKy503_Fdbs6cLq0cYXUqxuWIBdPjI/edit#

![image](https://user-images.githubusercontent.com/42398485/80011390-6ce08800-8491-11ea-859b-bb77dd3e08c2.png)

5. After removing a few digits from the `Business Phone` (+1 7138888888) >> +1 7138888 and tap on any empty place won't show the validation message **"Invalid business phone"**

![image](https://user-images.githubusercontent.com/42398485/80011759-eaa49380-8491-11ea-926d-563f0807aac3.png)

6. Some country texts are hidden.

![image](https://user-images.githubusercontent.com/42398485/80012094-68689f00-8492-11ea-80dd-36723eeebf4f.png)

**NOTE: Topcoder and Copilot flow testing is pending**

**Comment by maxceem** [comment]
> 1. `Company Name` is missing on the form.

Great, catch. - Fixed.✅

> 2. Since we have validations activated Title, Business Email and Business Phone labels are not in-line with the text fields

I was existent issue from the User Profile Settings page, but it's fixed now. ✅

> 3. Send My Request button is enabled even though the user filled an invalid email to the Business Email field

Fixed ✅

> 4. Send my Request button text should be Send My Request button

Fixed ✅

> 5. After removing a few digits from the Business Phone (+1 7138888888) >> +1 7138888 and tap on any empty place won't show the validation message "Invalid business phone"

Phone validation indeed works strange: ❌ will fix it

![phone-validation](https://user-images.githubusercontent.com/146016/80112225-8b29af00-8589-11ea-92a5-4b1de458d3b1.gif)

> 6. Some country texts are hidden.

❌ will fix it

**Comment by maxceem** [comment]
> 7. I guess we can enter white spaces for the required fields because I can see the Send my Request button is enabled.

![image](https://user-images.githubusercontent.com/146016/80115043-f32dc480-858c-11ea-91ef-aff0560caa6b.png)

⚠️ This is existent behaviour for most our forms, we treat multiple spaces as if value is entered. @vikasrohit should we change it and not allowed spaces as a value?

**Comment by maxceem** [comment]
A couple clarifications, @vikasrohit.

NOTE: As I understand the main task is to support users who registers using general Topcoder Registration form https://accounts.topcoder-dev.com/member/registration. FYI it only creates `basic_info` trait with the next fields:
```js
{
    "handle": "maxceem18",
    "firstName": "Max",
    "lastName": "18",
    "userId": 88771387,
    "status": "ACTIVE",
    "email": "email@gmail.com"
}
```

1. For customers users we might as "business email". As `connect_info` is not populated during initial registration, the business email would be empty and we can enter any email. As a result user might 2 different emails in 2 traits:

    ![image](https://user-images.githubusercontent.com/146016/80123274-50c70e80-8597-11ea-833a-6ca17eb1b0d2.png)

   Is it OK?    

2. Imagine if we register new user using general topcoder registration form. And and after we add some topcoder role so now if such a user creates a new project it would have Topcoder User flow during progressive registration, and would see the next form:

   ![image](https://user-images.githubusercontent.com/146016/80123764-02fed600-8598-11ea-84cf-3baddb397ec8.png)

    - we fill this form and send

    - server returns error:
       
       ```
       [The title should be provided, The company name should be provided]
       ```

    - So **Job Title** and **Company Name** are required in `connect_info` trait, but we don't ask about them in the form. 

    Should we add **Job Title** and **Company Name** as required fields for Topcoder flow?

**Comment by vigneshTheDev** [comment]
@maxceem 
> 5. After removing a few digits from the Business Phone (+1 7138888888) >> +1 7138888 and tap on any empty place won't show the validation message "Invalid business phone"

I see two kinds of validations for phone number input. 1. Earlier we used to rely on libphonenumber-js library (https://www.npmjs.com/package/libphonenumber-js). 2. And then according to this issue https://github.com/appirio-tech/accounts-app/issues/333#issuecomment-571023326, we've added a Regex based validation same as it's done in member service. 
These two ways of validations are mixed up that result in the strange behaviour of business phone input.

We can choose to handle this in two ways:
1. Rely on the regex based validation solely.

*Problems:*
* The country selection based on phone number selection still relies on libphonenumber-js.
So, when the phone number is not valid according to libphonenumber-js, country cannot be automatically selected.
* It's possible that the user submits non matching country and phone number.
![Peek 2020-04-24 02-25](https://user-images.githubusercontent.com/12785973/80148555-d9af6b80-85d2-11ea-95ea-e37ce665226f.gif)


2. Keep the libphonenumber-js based validation along with regex based validation (Recommended)

* In such a case, UI validation will not be same as member service but it'll be a more stricter validation while still making sure the phone number value is acceptable by the member service.
* The validation done with libphonenumber-js is more accurate. For example, +1 713888 is valid according to the regex match. But it's actually invalid (Check here: https://numverify.com).


Please let me know what you think.

**Comment by maxceem** [comment]
@vigneshTheDev thanks for detailed description.

We would prefer to use the 2nd *recommended* approach: keep both validations at the same time so it would be stricter than backend.

**Comment by vigneshTheDev** [comment]
Hi @maxceem, here's the pull request: https://github.com/appirio-tech/connect-app/pull/3961, please review

**Comment by maxceem** [comment]
Thank you @vigneshTheDev I'll review on Monday.

**Comment by maxceem** [comment]
@vikasrohit here is these questions left before we can pass it to QA Team https://github.com/appirio-tech/connect-app/issues/3833#issuecomment-618450070.

**Comment by vikasrohit** [comment]
@maxceem I don't think we should allow only white spaces in any field. I am not sure why it is the case with existing forms, but I never thought of having that even because of technical limitations.

**Comment by vikasrohit** [comment]
>For customers users we might as "business email". As connect_info is not populated during initial registration, the business email would be empty and we can enter any email. As a result user might 2 different emails in 2 traits:

I am not sure why it should happen if we are following the same approach as we have it right now in connect profile page. I mean, I just tested that for user who was registered with member registration or in any way it doesn't have connect_info trait. I was able see the same email, first name and last name (which are the only common fields between the two traits), which are there in `basic_info` trait, on the profile form. So should be the case here, isn't it?

> Should we add Job Title and Company Name as required fields for Topcoder flow?

Lets send these fields in hard coded way. Map Title to the User's role (if possible pick the highest privileged user role) and Company Name to Topcoder.

**Comment by maxceem** [comment]
> I am not sure why it should happen if we are following the same approach as we have it right now in connect profile page. I mean, I just tested that for user who was registered with member registration or in any way it doesn't have connect_info trait. I was able see the same email, first name and last name (which are the only common fields between the two traits), which are there in basic_info trait, on the profile form. So should be the case here, isn't it?

On the **My Profile** page we don't show **Business Email** field:
![image](https://user-images.githubusercontent.com/146016/80368120-e7941380-8894-11ea-9898-306859b701c4.png)

**Comment by vikasrohit** [comment]
> In our form we just have special logic to take them from basic_info and prepopulate these fields, if there is connect_info trait yet.

Lets use this in progressive form as well. I think that solve other concerns as well if update the basic_info trait as well when we are saving connect_info trait (I think we are also doing this on profile page right now).

**Comment by maxceem** [comment]
> Lets use this in progressive form as well. I think that solve other concerns as well if update the basic_info trait as well when we are saving connect_info trait (I think we are also doing this on profile page right now).

We already reuse this logic here. So we don't have issues with First/Last name.

The only thing I have concerns about is **Business Email**. We have it in Progressive Registration form, but we don't have it in the profile settings.

**Comment by vikasrohit** [comment]
Got it. We can do that same now for `Business Email` in progressive form as well i.e. read from `basic_info` if `connect_info` trait not available and when saving the progressive registration form, update both traits.

**Comment by maxceem** [comment]
> Got it. We can do that same now for Business Email in progressive form as well i.e. read from basic_info if connect_info trait not available and when saving the progressive registration form, update both traits.

Sure. Should users be allowed just edit it? As on the accounts page we have a special workflow for changing email with a confirmation email:

![image](https://user-images.githubusercontent.com/146016/80373546-e61b1900-889d-11ea-8c5c-3d36a12c163b.png)

**Comment by vikasrohit** [comment]
I think it should not be editable, if the email is already filled, because we still have the email change feature disabled in Connect.

**Comment by maxceem** [comment]
Topcoder members indeed cannot edit email in Connect. But customer user can edit email in Connect. But there is a special workflow with email confirmation.

**Comment by vikasrohit** [comment]
Lets not allow anyone to edit emails from the progressive registration form. Customers can still go to the regular Accounts and Security page to do that.

**Comment by maxceem** [comment]
@codejamtc this is now ready for QA.

Note: that as per https://github.com/appirio-tech/connect-app/issues/3833#issuecomment-620007826 we don't show **Business Email** for customers anymore.

**Comment by vikasrohit** [comment]
> we don't show Business Email for customers anymore.

@maxceem I meant the field to be there bot not editable. Is it possible to do that now?

**Comment by maxceem** [comment]
> @maxceem I meant the field to be there bot not editable. Is it possible to do that now?

Yes, will update.

**Comment by maxceem** [comment]
> @maxceem I meant the field to be there bot not editable. Is it possible to do that now?

Done, and update comment https://github.com/appirio-tech/connect-app/issues/3833#issuecomment-620488193

**Comment by codejamtc** [comment]
@maxceem @vikasrohit 

Issue 01: 🆗 
Issue 02: 🆗 
Issue 03: 🆗 
Issue 04: 🆗 
Issue 05: 🆗 
Issue 06: 🆗 
Issue 07: 🆗 
Issue 08: 🆗 https://github.com/appirio-tech/connect-app/issues/3833#issuecomment-620007826 - Lets not allow anyone to edit emails from the progressive registration form. Customers can still go to the regular Accounts and Security page to do that.

https://monosnap.com/file/Wj7ZqdVpyoLUps2jwNZKMexuGZkAkp

## Suggestions:

**Suggestions 01:** Better to change the label as `Job Title` rather than `Title`. 

![image](https://user-images.githubusercontent.com/42398485/80524528-5fc20e00-8955-11ea-9571-e97bd1954ecf.png)

**Suggestions 02:**  `Company name` to `Company Name`

![image](https://user-images.githubusercontent.com/42398485/80524626-884a0800-8955-11ea-882a-588019444d7e.png)

**Suggestions 03:**  No validation when `Start Time` >`End Time`

![image](https://user-images.githubusercontent.com/42398485/80524721-b2032f00-8955-11ea-9b9c-4a4b973c2ef8.png)

**Suggestions 04:**  `Local Timezone` and `Normal Working Hours` are optional so if user selects any values for that fields user can reset the data without restarting the form again. Should have an option to reset it. (Should be able to delete the current selection)

![image](https://user-images.githubusercontent.com/42398485/80525083-4a99af00-8956-11ea-9f2c-d7b30811cf07.png)

NOTE: Above are suggestions so these things won't directly affect the ticket so I will mark it as `Pass`

**Comment by maxceem** [comment]
Thanks @codejamtc for always checking on other things around. Your suggestions look very reasonable so I moved them to a separate issue for the product team to decide on them https://github.com/appirio-tech/connect-app/issues/3971

There is only one suggestion which I think we shouldn't implement.

> **Suggestions 03:**  No validation when `Start Time` >`End Time`

> ![image](https://user-images.githubusercontent.com/42398485/80524721-b2032f00-8955-11ea-9b9c-4a4b973c2ef8.png)

We support `Start Time` >`End Time`. This means that user works starting in the morning and works the whole day until the late night. For example here is a realistic example: 
   -  `Start Time`: `9:00` 
   -  `End Time`: `1:00`
   
   `Start Time` > `End Time` - it means that user starts an 9 am in the morning and works until 1 am late night.

**Comment by codejamtc** [comment]
@maxceem Agree, make sense 👍

**Comment by coderhacker** [comment]
Verified on production and this is working as expected. 

PS: Please create a new user from this form https://accounts.topcoder.com/member/registration
and follow steps in the attached video file.

[3833_prod_pass.zip](https://github.com/appirio-tech/connect-app/files/4557783/3833_prod_pass.zip)

CC @lakshmiathreya @drasticdpk
